### PR TITLE
Add cumulative operations: cumprod, cummin, and cummax

### PR DESCRIPTION
**Pull request by antimora**
Introduces three cumulative tensor operations (`cumprod`, `cummin`, and `cummax`) for both float and int tensors across all supported backends. These operations compute cumulative products, minimums, and maximums along a specified dimension, similar to PyTorch's corresponding operations.

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

**Issues:**
- Fixes https://github.com/tracel-ai/burn/issues/3818

**Related PRs:**
- https://github.com/tracel-ai/burn/pull/3806
- https://github.com/tracel-ai/burn/pull/3820
- https://github.com/tracel-ai/burn/pull/3821
- https://github.com/tracel-ai/cubecl/pull/863

### API
Added three new methods to `Tensor<B, D>` for all numeric kinds:
- `cumprod(dim: usize)` - Computes cumulative product along the specified dimension
- `cummin(dim: usize)` - Computes cumulative minimum along the specified dimension
- `cummax(dim: usize)` - Computes cumulative maximum along the specified dimension

All operations return a tensor with the same shape as the input.

### Backend Implementations

All three operations are implemented across all backends:

1. **ndarray**:
   - `cumprod`: Manual implementation using iterative multiplication with `ndarray::Zip`
   - `cummin`/`cummax`: Iterative comparison kernels

2. **tch (LibTorch)**: Native PyTorch bindings
   - `cumprod`: Direct binding to `tensor.cumprod(dim, dtype)`
   - `cummin`/`cummax`: Native PyTorch cummin/cummax operations

3. **candle**: Manual implementations
   - `cumprod`: Slice-and-concatenate approach (candle 0.9.1 lacks native cumprod)
   - `cummin`/`cummax`: Manual implementation using `broadcast_minimum`/`broadcast_maximum`
   - Handles both float and integer types with dtype conversion

4. **cubecl/wgpu (GPU)**: Custom kernel implementations
   - Naive sequential kernels (O(n²) complexity)
   - Suitable for small tensors; documented for future optimization with parallel scan algorithms
   - Similar approach to existing `cumsum` kernel

5. **autodiff**:
   - `cumprod`: **Full backward pass implementation**
     - Gradient formula: `(grad .* output).flip().cumsum().flip() / input`
     - Verified against PyTorch gradient computation
     - Stores input tensor for gradient calculation
   - `cummin`/`cummax`: **Panics with clear error message**
     - Proper gradient computation requires `scatter_add` (not yet implemented)
     - Documented limitation in API docs

6. **fusion/router**: Complete pass-through implementations
   - Proper IR integration for all three operations
   - Operation registration and routing for float and int tensors

### Intermediate Representation (IR)

- Added `CumProd(DimOpIr)`, `CumMin(DimOpIr)`, and `CumMax(DimOpIr)` to `BaseOperationIr` enum
- Uses `DimOpIr` (not `ReduceDimOpIr`) since output shape equals input shape
- Updated all pattern matches in fusion and router backends

### Forward Pass Tests
- **cumprod**: 5 comprehensive tests in `burn-tensor/src/tests/ops/cumprod.rs`
  - Coverage: Float/Int tensors, 2D/3D shapes, multiple dimensions
- **cummin**: 5 comprehensive tests in `burn-tensor/src/tests/ops/cummin.rs`
- **cummax**: Comprehensive test coverage in `burn-tensor/src/tests/ops/cummax.rs`
- All tests passing 

### Autodiff Gradient Tests
- **cumprod**: 2 gradient tests in `burn-autodiff/src/tests/cumprod.rs`
- **cummin/cummax**: Not applicable (operations panic in autodiff backend with clear error messages)

## Notes

- **Autodiff Limitation**: Only `cumprod` has full autodiff support. `cummin` and `cummax` are **not supported in the autodiff backend**
 and will panic with informative error messages. Proper gradient computation for min/max operations requires `scatter_add`, which is not
 yet implemented. This limitation is clearly documented in the API documentation.

- **Performance**: CubeCL implementations use naive O(n²) kernels suitable for small tensors. Future optimization with parallel scan
algorithms is planned and documented as TODOs.

### CODE REVIEW

File: crates/burn-autodiff/src/ops/tensor.rs [hunk 1]
```diff
@@ -1665,6 +1665,47 @@ impl<B: Backend, C: CheckpointStrategy> FloatTensorOps<Self> for Autodiff<B, C>
...
+                let output = B::float_cumprod(input.clone(), dim);
+
+                unary::<B, _>(ops.parents, ops.node, grads, |grad| {
+                    // Gradient of cumprod: grad_input[i] = sum(grad_output[j] * output[j] / input[i]) for all j >= i
+                    // This can be computed as: (grad * output).flip().cumsum().flip() / input
+                    let grad_times_output = B::float_mul(grad, output);
+                    let grad_reversed = B::float_flip(grad_times_output, &[dim]);
+                    let grad_cumsum = B::float_cumsum(grad_reversed, dim);
+                    let grad_flipped = B::float_flip(grad_cumsum, &[dim]);
+                    B::float_div(grad_flipped, input)
```
Comment (laggui) [inline crates/burn-autodiff/src/ops/tensor.rs]: Would be nice to avoid the double flipping (potential improvement), even if with fusion this could be mitigated.

File: crates/burn-candle/src/ops/int_tensor.rs [hunk 2]
```diff
@@ -303,6 +303,25 @@ impl<F: FloatCandleElement, I: IntCandleElement> IntTensorOps<Self> for Candle<F
...
+        slices.push(tensor_float.narrow(dim, 0, 1).unwrap());
+
+        for i in 1..dim_size {
+            let curr = tensor_float.narrow(dim, i, 1).unwrap();
+            let prod = slices[i - 1].broadcast_mul(&curr).unwrap();
+            slices.push(prod);
+        }
+
+        let result_float = candle_core::Tensor::cat(&slices, dim).unwrap();
+        CandleTensor::new(result_float.to_dtype(dtype).unwrap())
```
Comment (laggui) [inline crates/burn-candle/src/ops/int_tensor.rs]: We should make this into a more generic implementation, so it can be called for float. And for other cumulative ops which only change the operator applied.

File: crates/burn-cubecl/src/ops/numeric.rs [hunk 3]
```diff
@@ -318,3 +318,86 @@ pub fn cumsum<R: CubeRuntime, E: CubeElement>(input: CubeTensor<R>, dim: usize)
...
     output
 }
+
+#[cube(launch)]
+fn cumprod_kernel<C: Numeric>(
+    input: &Tensor<C>,
+    output: &mut Tensor<C>,
+    dim_stride: u32,
+    #[comptime] dim_size: u32,
+) {
```
Comment (laggui) [inline crates/burn-cubecl/src/ops/numeric.rs]: Same with the kernel and launch. The instruction/operation can be provided as a config, similar to the reduce family of kernels. Or even just the comparison ops family.

File: crates/burn-ndarray/src/ops/macros.rs [hunk 4]
```diff
@@ -68,3 +68,21 @@ pub(crate) fn cumsum_dim<E: NdArrayElement>(tensor: SharedArray<E>, dim: usize)
...
+    let dim_size = shape[dim];
+    for i in 1..dim_size {
+        let prev = result.index_axis(axis, i - 1).to_owned();
+        let mut current = result.index_axis_mut(axis, i);
+        Zip::from(&mut current)
+            .and(&prev)
+            .for_each(|c, &p| *c = c.mul(p.elem()));
+    }
+
+    result.into_shared()
```
Comment (laggui) [inline crates/burn-ndarray/src/ops/macros.rs]: Same for ndarray

File: crates/burn-autodiff/src/ops/tensor.rs (hunk 1 above)
Comment (antimora) [reply crates/burn-autodiff/src/ops/tensor.rs]: Optimize cumprod gradient with negative step slicing, which should be faster considering slicing is just stride manipulation.

**Comment by antimora** [comment]
@laggui added autodiff for cummin and cummax

File: crates/burn-candle/src/ops/int_tensor.rs [hunk 5]
```diff
@@ -303,6 +303,53 @@ impl<F: FloatCandleElement, I: IntCandleElement> IntTensorOps<Self> for Candle<F
...
+
+        let result_float = cumulative_op!(
+            tensor_float,
+            dim,
+            tensor_float.narrow(dim, 0, 1).unwrap(),
+            |prev: &candle_core::Tensor, curr: &candle_core::Tensor| {
+                prev.broadcast_mul(curr).unwrap()
+            }
+        );
+        CandleTensor::new(result_float.to_dtype(dtype).unwrap())
```
Comment (laggui) [inline crates/burn-candle/src/ops/int_tensor.rs]: We don't need to use a macro for this, just define a cumulative function that takes a closure for the op:

```rust
pub fn cumulative_with_op<F>(tensor: &candle_core::Tensor, dim: usize, op: F) -> candle_core::Tensor
where
    F: Fn(&candle_core::Tensor, &candle_core::Tensor) -> candle_core::Result<candle_core::Tensor>,
{
    let dim_size = tensor.dims()[dim];
    let mut slices = Vec::with_capacity(dim_size);

    // First slice is the initial value
    slices.push(tensor.narrow(dim, 0, 1).unwrap());

    // Apply cumulative operation
    for i in 1..dim_size {
        let curr = tensor.narrow(dim, i, 1).unwrap();
        let result = op(&slices[i - 1], &curr).unwrap();
        slices.push(result);
    }

    candle_core::Tensor::cat(&slices, dim).unwrap()
}
```

The resulting call will be much simpler:
```rust
let result = cumulative_with_op(&tensor.tensor, dim, |prev, curr| prev.broadcast_mul(curr));

CandleTensor::new(result)
```

File: crates/burn-cubecl/src/ops/numeric.rs [hunk 6]
```diff
@@ -265,56 +269,160 @@ fn cumsum_kernel<C: Numeric>(
...
+        }
+    }
+    output[idx] = max_val;
+}
 
-    output
+/// Macro to generate cumulative operation wrapper functions
+///
+/// This reduces duplication across cumsum, cumprod, cummin, cummax implementations.
+macro_rules! cumulative_op {
```
Comment (laggui) [inline crates/burn-cubecl/src/ops/numeric.rs]: We also don't need/want a macro for the cubecl kernels. It is preferable to implement this as an operation family (e.g., reduce, comparison or even binary op family as per my previous comment).

More concretely:
```rust

pub(crate) trait CumulativeOpFamily: Send + Sync + 'static {
    type CumulativeOp<C: Numeric>: CumulativeOp<C>;
}

#[cube]
pub(crate) trait CumulativeOp<C: Numeric>: 'static + Send + Sync {
    /// Execute a cumulative operation.
    fn execute(lhs: C, rhs: C) -> C;
}

struct SumOp;
struct ProdOp;
struct MaxOp;
struct MinOp;

impl CumulativeOpFamily for SumOp {
    type CumulativeOp<C: Numeric> = Self;
}

impl CumulativeOpFamily for ProdOp {
    type CumulativeOp<C: Numeric> = Self;
}

impl CumulativeOpFamily for MaxOp {
    type CumulativeOp<C: Numeric> = Self;
}

impl CumulativeOpFamily for MinOp {
    type CumulativeOp<C: Numeric> = Self;
}

#[cube]
impl<N: Numeric> CumulativeOp<N> for SumOp {
    fn execute(lhs: N, rhs: N) -> N {
        lhs + rhs
    }
}

#[cube]
impl<N: Numeric> CumulativeOp<N> for ProdOp {
    fn execute(lhs: N, rhs: N) -> N {
        lhs * rhs
    }
}

#[cube]
impl<N: Numeric> CumulativeOp<N> for MaxOp {
    fn execute(lhs: N, rhs: N) -> N {
        N::max(lhs, rhs)
    }
}

#[cube]
impl<N: Numeric> CumulativeOp<N> for MinOp {
    fn execute(lhs: N, rhs: N) -> N {
        N::min(lhs, rhs)
    }
}
```

You'll be able to define a single kernel for this family, generic over the op:
```rust
#[cube(launch)]
fn cumulative_kernel<C: Numeric, O: CumulativeOpFamily>
```
with something like:
```rust
result = O::CumulativeOp::<C>::execute(result, input[read_idx]);
```

Now, the only think that changes is the initial condition for sum/prod/min/max. But given your current implementation, this could probably just be provided as a comptime value (0, 1, `C::MAX` and `C::MIN` respectively).

File: crates/burn-ndarray/src/ops/macros.rs [hunk 7]
```diff
@@ -51,20 +51,42 @@ pub(crate) fn prod_dim<E: NdArrayElement>(tensor: SharedArray<E>, dim: usize) ->
...
 
-pub(crate) fn cumsum_dim<E: NdArrayElement>(tensor: SharedArray<E>, dim: usize) -> SharedArray<E> {
-    let axis = Axis(dim);
-    let shape = tensor.shape().to_vec();
-    let mut result = tensor.to_owned();
+/// Macro to generate cumulative operation functions with minimal code duplication.
+///
+/// Generates the boilerplate for cumulative operations (cumsum, cumprod, cummin, cummax)
+/// by only requiring the operation-specific logic to be specified.
+macro_rules! cumulative_fn {
```
Comment (laggui) [inline crates/burn-ndarray/src/ops/macros.rs]: Can probably change this to a function with closure, similar to the candle suggestion

File: crates/burn-cubecl/src/ops/numeric.rs [hunk 8]
```diff
@@ -246,63 +330,115 @@ fn cumsum_kernel<C: Numeric>(
...
+///
+/// This is a **naive sequential implementation** along the product dimension:
+/// - Each output element sequentially reads all previous elements along the dimension
+/// - Computational complexity: O(n²) memory reads where n is the size of the product dimension
+/// - **Performance:** Suitable for small tensors or small dimensions. For large tensors,
+///   performance will degrade significantly compared to an optimized parallel scan algorithm.
+///
+/// # TODO
+///
+/// Implement an efficient GPU-optimized parallel scan algorithm.
```
Comment (laggui) [inline crates/burn-cubecl/src/ops/numeric.rs]: This generally applies to all cumulative kernels here that re-use the same implementation. We can remove this repeated note here and at every other definition. Keep it linked to the actual kernel implementation.

File: crates/burn-ndarray/src/ops/macros.rs [hunk 9]
```diff
@@ -51,20 +51,56 @@ pub(crate) fn prod_dim<E: NdArrayElement>(tensor: SharedArray<E>, dim: usize) ->
         .into_shared()
 }
 
-pub(crate) fn cumsum_dim<E: NdArrayElement>(tensor: SharedArray<E>, dim: usize) -> SharedArray<E> {
+/// Generic cumulative operation function with closure-based operation.
+///
+/// Reduces code duplication for cumulative operations (cumsum, cumprod, cummin, cummax)
+/// by accepting a closure that defines the operation-specific logic.
```
Comment (laggui) [inline crates/burn-ndarray/src/ops/macros.rs]: Not relevant. The function signature and `op` arg documentation is clear enough.

File: crates/burn-tensor/src/tensor/api/numeric.rs [hunk 10]
```diff
@@ -695,6 +695,100 @@ where
...
+    /// Computes the cumulative minimum of elements along the given *dimension* or *axis*.
+    ///
+    /// # Arguments
+    ///
+    /// * `dim` - The dimension or axis along which to compute the cumulative minimum.
+    ///
+    /// # Note
+    ///
+    /// This operation is **not supported for the autodiff backend** and will panic.
+    /// Proper gradient computation requires scatter_add which is not yet implemented.
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/numeric.rs]: Forgot to update the doc, this is not accurate

File: crates/burn-tensor/src/tensor/api/numeric.rs [hunk 11]
```diff
@@ -695,6 +695,100 @@ where
...
+    /// Computes the cumulative maximum of elements along the given *dimension* or *axis*.
+    ///
+    /// # Arguments
+    ///
+    /// * `dim` - The dimension or axis along which to compute the cumulative maximum.
+    ///
+    /// # Note
+    ///
+    /// This operation is **not supported for the autodiff backend** and will panic.
+    /// Proper gradient computation requires scatter_add which is not yet implemented.
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/numeric.rs]: Same

File: crates/burn-tensor/src/tensor/api/numeric.rs [hunk 12]
```diff
@@ -2794,6 +2888,61 @@ where
...
+    /// Computes the cumulative minimum of elements along a dimension.
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The tensor to compute the cumulative minimum of.
+    /// * `dim` - The dimension along which to compute the cumulative minimum.
+    ///
+    /// # Returns
+    ///
+    /// A tensor with the same shape as the input tensor, where each element is the minimum    /// of all elements up to and including that position along the specified dimension.
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/numeric.rs]: Looks like you merged cumprod/cummin docs :smile:

File: crates/burn-autodiff/src/ops/tensor.rs [hunk 13]
```diff
@@ -1654,6 +1654,185 @@ impl<B: Backend, C: CheckpointStrategy> FloatTensorOps<Self> for Autodiff<B, C>
...
+                    let shape = grad_times_output.shape();
+                    let mut slices = vec![burn_tensor::Slice::full(); shape.num_dims()];
+                    slices[dim] = burn_tensor::Slice::with_step(0, None, -1);
+
+                    // Reverse, cumsum, reverse back using negative step slicing
+                    let grad_reversed = B::float_slice(grad_times_output, &slices);
+                    let grad_cumsum = B::float_cumsum(grad_reversed, dim);
+                    let grad_result = B::float_slice(grad_cumsum, &slices);
+
+                    B::float_div(grad_result, input)
```
Comment (laggui) [inline crates/burn-autodiff/src/ops/tensor.rs:1697]: Wouldn't this lead to NaNs when there are zeros in the input?

We should add a test with such edge cases that require more careful handling.

Comment (antimora) [reply crates/burn-autodiff/src/ops/tensor.rs:1697]: Yes, you're correct. The current cumprod gradient implementation uses division (grad / input), which produces NaN
when the input contains zeros. Deeper research into this tells me this is a well-known challenging problem across multiple deep learning frameworks.

I have added tests and make a note. We need more sophisticated algorithm. For now we can ask users to handle zeros. Even Pytorch has some cases when it returns NaN: https://discuss.pytorch.org/t/error-during-backward-pass-function-cumprodbackward0-returned-nan-values-in-its-0th-output/205018

I have filed issue to come back to: https://github.com/tracel-ai/burn/issues/3864 

We need better Burn primitive ops.

**Review by laggui — APPROVED** [review]
